man/meteolabel.1
.TH METEOLABEL "1" "July 2003" "Meteo station tools" Othello
.SH NAME
meteolabel \- compute labels to identify graphs to meteodraw
.SH SYNOPSIS
meteolabel [
.B \-s
] [
.B \-c
] [
.B \-L
]
.I label
[
.I operator ...
]
.SH DESCRIPTION
The 
.BR meteolabel (1)
programm computes labels that specify the interval and end time for
a graph.
This program is mainly used inside the meteobrowser PHP script.

Without operator arguments, the 
.BR meteolabel (1)
parses the
.I label
and diplays it in a format specified by the options. By default, it
will simply format it as a normalized label.

The operator arguments allow to navigate between different levels
of resolution and in time. The following operators are known:
.TP
up
move up one level from the current label. E.g. the day label d20030709
becomes
.TP
upup
move up two levels. Note that moving up from d20021231 one level gives
week 0 in 2003, so that moving up again gives january 2003, or the 
label m200301. Applying the upup operator to d20021231 gives december
2003, i.e. the label m200212.
.TP
upupup
move up three levels.
.TP
down
move down one level from the current label. E.g. the year label y2003
becomes the month label m200307.
.TP
downdown
move down two levels, see the description of upup for an explantion
why this is different from moving down twice.
.TP
downdowndown
move down three levels.
.TP
previous
move forward in time, e.g. the week label w200314 becomes w200315.
.TP
next
move backward in time, e.g. the month label m200301 becomes m200212


.SH OPTIONS
.TP
.B \-c
display the central time of the current label in 
.BR ctime (3)
format.
.TP
.B \-s
display the central time of the current label as Unix time in seconds
since midnight of 1st januar 1970.
.TP
.B \-L
suppress the default display of the label.

.SH "SEE ALSO"
.BR meteo (1)

.SH AUTHOR
Dr. Andreas Mueller <afm@othello.ch>


NEWS
0.9.17	a new outlet interface can be used to process meteo station data
	directly. In contrast to the map interface, the data is pushed
	to a file or into a pipe by the meteopoll process every minute.

0.9.15	new interface which allows to read current readings from the console
	to display on the web page. The new -m option to meteopoll makes
	sure current data is written to a file (the argument to the option)
	in XML format. Since updates to this file are very frequent, one
	should read the file atomically, one should use the meteomap program
	to read the data.
	fix of a nasty bug in the week label code, causing navigation failure
	at week resolution in some cases

0.9.14	gallery files included as samples
	can now compute sunrise and sunset (through meteosunriset and Sun.cc)

0.9.13	minor bug in meteobrowser.php
	gallery pages

0.9.12	some minor fixes, plus a security fix: remote code execution in
	meteobrowser.php because of missing escapeshellcmd

0.9.11	some bug fixes for GroWeather

0.9.10	turn of CRC0 for GroWeather and possibly other type E weatherlinks.

0.9.9	A hack to protect from IE reading the complete meteobrowser.php
	space offline

0.9.8	fixes a integer overflow bug: to compute the time labels for the
	maps, GraphWindow.cc used the formular (start + end)/2, but since
	about the middle of january, this causes an overflow, so that the
	target ends up in february 1936. Replace that by start/2 + end/2.
	
0.9.7	some bug fixes
	removal of offset configuration variable
	added averages configuration in XML file, so the user now gets full
	control about which averages are compute, and how
	battery and transmitter fields now work
	sunrise and sunset can also be stored in the database (this is a
		waste of space, but who cares about a few gibabytes these
		days)
	added channel of type sun, which indicates the time during which
		the sun is above the horizon

0.9.6	some bug fixes
	removal of header table

0.9.5	some minor fixes

(for further news please check the ChangeLog)

0.9.0	Major rewrite of many aspects of meteo sponsored by the University
	Univates in Lajeado in Rio Grande do Sul, Southern Brazil. Further
	contributions: Jeff Wabik (new derived value computations).

--
$Id: NEWS,v 1.32 2009/01/10 19:00:23 afm Exp $


src/meteolabel.cc
/*
 * meteolabel.cc -- utility program to convert timestamps into labels
 *                  and navigating up and down
 *
 * (c) 2003 Dr. Andreas Mueller, Beratung und Entwicklung
 *
 * $Id: meteolabel.cc,v 1.6 2004/02/25 23:41:13 afm Exp $
 */
#include <Timelabel.h>
#include <MeteoException.h>
#include <iostream>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int	meteolabel(int argc, char *argv[]) {
	int	c;
	bool	outtime = false, outctime = false, outlabel = true,
		outtitle = false;

	// read the command line
	while (EOF != (c = getopt(argc, argv, "scLt")))
		switch (c) {
		case 's':
			// output as timestamp
			outtime = true;
			break;
		case 'c':
			// output ctime string
			outctime = true;
			break;
		case 't':
			// output as string suitable for a title
			outtitle = true;
			break;
		case 'L':
			// output as a label
			outlabel = false;
			break;
		}

	// there must be at least one more arguments
	if (optind >= argc) {
		throw meteo::MeteoException("must specify label argument", "");
	}

	// next argument is a label
	std::string		label = argv[optind++];
	meteo::Timelabel	tl;
	try {
		tl = meteo::Timelabel(label);
	} catch (...) {
		std::cerr << "cannot convert label '" << label << "'"
			<< std::endl;
		exit(EXIT_FAILURE);
	}

	// following arguments are operators
	for (int i = optind; i < argc; i++) {
		if (!strcmp(argv[i], "up")) {
			tl = tl.up();
		}
		if (!strcmp(argv[i], "upup")) {
			tl = tl.upup();
		}
		if (!strcmp(argv[i], "upupup")) {
			tl = tl.upupup();
		}
		if (!strcmp(argv[i], "down")) {
			tl = tl.down();
		}
		if (!strcmp(argv[i], "downdown")) {
			tl = tl.downdown();
		}
		if (!strcmp(argv[i], "downdowndown")) {
			tl = tl.downdowndown();
		}
		if (!strcmp(argv[i], "previous")) {
			tl = tl.previous();
		}
		if (!strcmp(argv[i], "next")) {
			tl = tl.next();
		}
	}

	// output in the desired format
	if (outtime)
		std::cout << tl.getTime() << std::endl;
	if (outctime)
		std::cout << tl.getCtime();	// no endl, ctime includes one
	if (outlabel)
		std::cout << tl.getString() << std::endl;
	if (outtitle)
		std::cout << tl.getTitle() << std::endl;

	exit(EXIT_SUCCESS);
}


int	main(int argc, char *argv[]) {
	try {
		meteolabel(argc, argv);
	} catch (meteo::MeteoException& me) {
		fprintf(stderr, "MeteoException in meteolabel: %s/%s\n",
			me.getReason().c_str(), me.getAddinfo().c_str());
		exit(EXIT_FAILURE);
	}
	exit(EXIT_SUCCESS);
}
